### PR TITLE
Policy related changes for Qubes 4.0

### PR DESCRIPTION
**Pull request by marmarek**
- add `$adminvm` keyword (QubesOS/qubes-issues#2872)
- allow spaces in action options (`allow, target=$adminvm`)
- generate default policy for Admin API calls based on methods annotations  (QubesOS/qubes-issues#2871)
-  tool to visualize qrexec policy (QubesOS/qubes-issues#2873)

This will conflict with #120, but the conflict will be easy to resolve and this way it will be easier to review.

Cc @rootkovska

### CODE REVIEW

**Comment by rootkovska** [comment]
On Mon, Jun 26, 2017 at 09:39:17PM -0700, Marek Marczykowski-Górecki wrote:
> - add `$adminvm` keyword (QubesOS/qubes-issues#2872)
> - allow spaces in action options (`allow, target=$adminvm`)
> - generate default policy for Admin API calls based on methods annotations  (QubesOS/qubes-issues#2871)
> -  tool to visualize qrexec policy (QubesOS/qubes-issues#2873)
O, a to mieliśmy dopiero w 4.1 zrobić?

> 
> This will conflict with #120, but the conflict will be easy to resolve and this way it will be easier to review.
> 

Do czego odnosi się #120?

j.

File: doc/manpages/qrexec-policy-graph.rst [hunk 1]
```diff
@@ -0,0 +1,73 @@
+.. program:: qrexec-policy-graph
+
+:program:`qrexec-policy-graph` -- Graph qrexec policy
+=====================================================
+
+Synopsis
+--------
+
+:command:`qrexec-policy-graph` skel-manpage.py [-h] [--include-ask] [--source *SOURCE* [*SOURCE* ...]] [--target *TARGET* [*TARGET* ...]] [--service *SERVICE* [*SERVICE* ...]] [--output *OUTPUT*] [--policy-dir POLICY_DIR] [--system-info SYSTEM_INFO]
```
Comment (woju) [inline doc/manpages/qrexec-policy-graph.rst]: `s/ skel-manpage\.py//`
This one is actually my fault, and long standing bug in the tool for generating manpage drafts.

File: qubes-rpc-policy/admin-global-ro [hunk 2]
```diff
@@ -0,0 +1,13 @@
+## This file is included from all global read-only admin.* policy files 
+## _in default configuration_. To allow only specific action,
```
Comment (woju) [inline qubes-rpc-policy/admin-global-ro]: trailing whitespace

File: qubes-rpc-policy/admin-global-rwx [hunk 3]
```diff
@@ -0,0 +1,11 @@
+## This file is included from all global read-write admin.* policy files 
+## _in default configuration_. To allow only specific action,
```
Comment (woju) [inline qubes-rpc-policy/admin-global-rwx]: trailing whitespace

File: qubes-rpc-policy/admin-local-ro [hunk 4]
```diff
@@ -0,0 +1,14 @@
+## This file is included from all local read-only admin.* policy files 
+## _in default configuration_. To allow only specific action, 
+## edit specific policy file.
```
Comment (woju) [inline qubes-rpc-policy/admin-local-ro]: trailing whitespace, possibly copy-pasted

File: qubes-rpc-policy/admin-local-rwx [hunk 5]
```diff
@@ -1,5 +1,6 @@
-## This file is included from all admin.* policy files _in default
-## configuration_. To allow only specific action, edit specific policy file.
+## This file is included from all local read-write admin.* policy files 
+## _in default configuration_. To allow only specific action,
```
Comment (woju) [inline qubes-rpc-policy/admin-local-rwx]: trailing whitespace, most likely copy-pasted

File: qubes-rpc-policy/generate-admin-policy [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+default_policy_header = '''\
+## Note that policy parsing stops at the first match.
+## Anything not specifically allowed here (or in included file) will be denied.
+
+## Please use a single # to start your custom comments
+
+## Add your entries here, make sure to append ",target=dom0" to all allow/ask actions
+
+## Include a common file for all admin.* methods to ease setting up
```
Comment (woju) [inline qubes-rpc-policy/generate-admin-policy]: trailing whitespace

File: qubespolicy/__init__.py [hunk 7]
```diff
@@ -121,11 +125,11 @@ def __init__(self, line, filename=None, lineno=None):
         self.filename = filename
 
         try:
-            self.source, self.target, self.full_action = line.split()
+            self.source, self.target, self.full_action = line.split(maxsplit=2)
         except ValueError:
             raise PolicySyntaxError(filename, lineno, 'wrong number of fields')
 
-        (action, *params) = self.full_action.split(',')
+        (action, *params) = self.full_action.replace(' ', '').split(',')
```
Comment (woju) [inline qubespolicy/__init__.py]: IIUC, this means that params can be additionally separated by spaces preceding or following the comma, but comma is still mandatory and tabs won't work (which is inconsistent to separators between source, target and full_action, where they are perfectly fine).

File: qubespolicy/graph.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    if args.target:
+        targets = args.target
+    else:
+        targets.append('$dispvm')
+        targets.extend('$dispvm:' + dom for dom in system_info['domains']
+            if system_info['domains'][dom]['dispvm_allowed'])
+
+    connections = set()
+
+    output.write('digraph g {\n')
```
Comment (woju) [inline qubespolicy/graph.py]: This one looks like prime candidate for writing in jinja. But at this point I don't think that would be appropriate. :)

File: qubespolicy/__init__.py (hunk 7 above)
Comment (marmarek) [inline qubespolicy/__init__.py]: arguments to the action are still separated by coma(s). The intention here is to tolerate spaces, as many would add after coma

**Pull request merged**